src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -141,157 +173,99 @@ internal void Init(NpgsqlCommand command, CommandBehavior behavior, List<NpgsqlS
-                catch (PostgresException)
-                {
-                    State = ReaderState.Consumed;
-                    throw;
-                }
-            }
+            return msg.Code == BackendMessageCode.DataRow;
         }
 
-        internal void ProcessMessage(IBackendMessage msg)
+        async Task<bool> Read(bool async)
         {
-            Debug.Assert(msg != null);
-
-            switch (msg.Code)
+            switch (State)
             {
-            case BackendMessageCode.DataRow:
-                Debug.Assert(RowDescription != null);
-                if (Connector.State != ConnectorState.Fetching)
-                    Connector.State = ConnectorState.Fetching;
-                ProcessDataMessage((DataRowMessage)msg);
-                _hasRows = true;
-                switch (State)
-                {
-                case ReaderState.BetweenResults:
-                    State = ReaderState.BeforeResult;
-                    break;
-                case ReaderState.BeforeResult:
-                    State = ReaderState.InResult;
-                    break;
-                case ReaderState.InResult:
-                    break;
-                default:
-                    throw Connector.UnexpectedMessageReceived(BackendMessageCode.DataRow);
-                }
-                return;
+            case ReaderState.BeforeResult:
+                // First Read() after NextResult. Data row has already been processed.
+                State = ReaderState.InResult;
+                return true;
 
-            case BackendMessageCode.CompletedResponse:
-                var completed = (CommandCompleteMessage) msg;
-                switch (completed.StatementType)
+            case ReaderState.InResult:
+                await ConsumeRow(async);
+                if ((_behavior & CommandBehavior.SingleRow) != 0)
                 {
-                case StatementType.Update:
-                case StatementType.Insert:
-                case StatementType.Delete:
-                case StatementType.Copy:
-                case StatementType.Move:
-                    if (!_recordsAffected.HasValue)
-                        _recordsAffected = 0;
-                    _recordsAffected += completed.Rows;
-                    break;
+                    // TODO: See optimization proposal in #410
+                    await Consume(async);
+                    return false;
                 }
+                break;
 
-                _statements[StatementIndex].ApplyCommandComplete(completed);
-                goto case BackendMessageCode.EmptyQueryResponse;
-
-            case BackendMessageCode.EmptyQueryResponse:
-                State = ReaderState.BetweenResults;
-                return;
+            case ReaderState.BetweenResults:
+            case ReaderState.Consumed:
+            case ReaderState.Closed:
+                return false;
+            default:
+                throw new ArgumentOutOfRangeException();
```

Comment:
I don't think `NotSupportedException` is the right one either - that's about the user trying to use a feature that isn't supported. What's happening here is about a programming error when we add a value the enum but don't add the case in the switch.

I looked some time ago for a nicer exception but couldn't fine one. Since it's about a programming error it doesn't really matter - it's a bit like a debug assertion (i.e. it should never happen).

If you feel very strongly about it, then we need to replace it everywhere, so maybe a different PR (although I don't think our time is worth it).

Reply:
You're right about `NotSupportedException`, but still wrong about `ArgumentOutOfRangeException`. The documentation says:

> You should throw an `InvalidOperationException` exception only when the state of your object for some reason does not support a particular method call. That is, the method call is valid in some circumstances or contexts, but is invalid in others. 
If the method invocation failure is due to invalid arguments, then `ArgumentException` or one of its derived classes, `ArgumentNullException` or `ArgumentOutOfRangeException`, should be thrown instead.

Okay, let's leave it as is for now.